examples/angularjs/js/controllers/todoCtrl.js

```diff
@@ -76,14 +88,31 @@ angular.module('todomvc')
 				$scope.editedTodo = null;
 				return;
 			}
-
+			// Saving the todo.
+			todo.media = null;
 			store[todo.title ? 'put' : 'delete'](todo)
-				.then(function success() {}, function error() {
-					todo.title = $scope.originalTodo.title;
-				})
-				.finally(function () {
-					$scope.editedTodo = null;
-				});
+			.then((todos) => {
```

Comment:
A few things here:
 
 * There should be one method to fetch and update todo item with gif image. This method should  then be reused for insert and update.
 * Do not set `media` to null. It's ok to keep it defined in case API is unreachable and a new gif cannot be fetched.
 * When working with promises it is advisable to keep the promise flow at one level. Nesting promises is bad style and leads to poorly readable code and bugs. 
 * One of the bugs is here - the `index` parameter passed to `store.put` is actually not used in the `put` method. It was used before but not anymore after your changes.
 * Dealing with immutable objects is better than mutating existing ones. This helps avoid bugs and makes the code easier to read. Languages like Scala can enforce this out of the box, but not JavaScript.
 * Some things here can be made a bit more readable with using lodash.

Something along the lines you want to have in your reusable function:

```javascript
                function todoItemGifUpdateFactory(item, itemIndex) {
			return function(gifUrl) {
                                // immutable update.
				var updatedItem = _.chain(item).clone().set('media', gifUrl).value();
				return store.put(updatedItem, itemIndex);
			}
		}
```

```javascript
                        // Saving the todo.
			store.insert(newTodo)
			.then(function (todos) {
				$scope.saving = false;
				return todos.length - 1;
			})
			.then(function(indexOfInsertedItem) {
				$scope.newTodo = '';
				return gifFetch.get(newTodo.title)
					.then(todoItemGifUpdateFactory(newTodo, indexOfInsertedItem))
			})
			.then(_.noop, function(err) {
				$scope.saving = false;
				console.warn('Could not update gif image', err);
			});
```